Network stack: net_if cleanup

tbursztyka opened the pull request:
Various commit to fix #8728 issue.

Commit:
net/iface: Reduce usage of ifdef around CONFIG_NET_IPV6

Let's regroup all the IPv6 related function into one ifdef. There is no
need to implement a dummy function for each: if IPv6 is not enabled,
these functions are unused in the core.

Fixes #8728

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/iface: Reduce usage of ifdef around CONFIG_NET_IPV4

Let's regroup all the IPv4 related function into one ifdef. There is no
need to implement a dummy function for each: if IPv4 is not enabled,
these functions are unused in the core.

Fixes #8728

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/iface: Reduce ifdef usage on various options

Offload, DAD, RS and IPv4 autoconf.

Fixes #8728

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/iface: Initialize IPv4/6 address so it removes more ifdef

Just create init functions for each IP version.

Fixes #8728

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/iface: Remove useless attributes in net_if_ipv6 structure

DAD use dad_count attribute in struct net_if_addr, since DAD is ran on
each and every ipv6 unicast address.

Fixes #8728

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/iface: Coalesce all DAD timers through one

This reduces the size of struct net_if_addr by 24 bytes by moving
the k_delayed_work attribute into net_if core code.
Then each net_if_addr can be added to the timer handler via a slist.

This does not make much gain if the system has only 1 unicast IPv6
address. It's a nice memory improvment once it has 2+ unicast IPv6
address. Note that having IPv4 enabled along with IPv6 will also see
memory improvements since both IPv6 and IPv4 use the same struct
net_if_addr.

Fixes #8728

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/iface: Coalesce all RS timers through one

This reduces the size of struct net_if_ipv6 by 24 bytes by moving
the k_delayed_work attribute into net_if core code.
Then each net_if_ipv6 can be added to the timer handler via a slist.

This does not make much gain if the system has only 1 network interface
It starts to be interesting if it has 2+ network interfaces then.

Fixes #8728

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv4: Add a net mgmt event for IPv4 router deletion

There will probably be at some point a way to remove IPv4 router, and
thus it will require such event.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/iface: Rework how routers are handled

- router lifetime is always a u16_t so fixing
net_if_ipv6_router_update_lifetime() signature.
- Coalescing router timers into one: this reduces the net_if_router
structure by 22 bytes
- refactor IPv6 and IPv4 router code so it's handled in generic
functions, to avoid duplicating 90% of the code for each family. This
also fixes the lifetime support for IPv4 which was missing.

Note however that IPv4 routing support seems to be missing as none of
the relevant functions are used anywhere yet.

Fixes #8728

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

jukkar commented on subsys/net/ip/net_if.c:
```diff
@@ -1986,9 +1986,9 @@ u32_t net_if_ipv6_calc_reachable_time(struct net_if_ipv6 *ipv6)
 #define join_mcast_nodes(...)
```
Typo in commit msg "s/IPv6/IPv4"
Also better imho "There is no need to implement a dummy function for each as if IPv4 is not enabled then these functions are unused in the core."

jukkar replied:
Now there is duplicate sentences in commit msg. I suggest we remove the "There is no need to implement a dummy function for each: if IPv4 is not enabled, these functions are unused in the core." as that was what I suggested in my comment. Same for the IPv6 patch.

tbursztyka replied:
indeed, c/p issue

jukkar commented on subsys/net/ip/net_if.c:
```diff
@@ -412,6 +414,221 @@ static u8_t get_ipaddr_diff(const u8_t *src, const u8_t *dst, int addr_len)
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&active_router_timers,
+					 router, next, node) {
+		new_timer = MIN(router->life_start +
```
Avoid arithmetic things in MIN() and MAX() macros, they might have hard to see interactions (I am not saying there is here)

Commit:
net/iface: Add IPv4 route find and add

Removing an IPv4 router was missing, as well as finding the default
router for an IPv4 address.

Note howevere that IPv4 router features are not used anywhere yet. But
at least the API is there and is a 1:1 to IPv6, if that matters.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

jukkar commented on subsys/net/ip/net_if.c:
```diff
@@ -404,6 +406,222 @@ static u8_t get_ipaddr_diff(const u8_t *src, const u8_t *dst, int addr_len)
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&active_router_timers,
+					  router, next, node) {
+
+		if (router->life_start +
```
It looks like this does not take wraparounds into account

tbursztyka commented:
recheck

tbursztyka commented on subsys/net/ip/net_if.c:
```diff
@@ -622,18 +888,33 @@ void net_if_ipv6_dad_failed(struct net_if *iface, const struct in6_addr *addr)
 #else
+void net_if_start_dad(struct net_if *iface)
+{
+	join_mcast_nodes(iface,
```
hum original was not nice either. I have a better solution

tbursztyka commented:
@jukkar can you request a recheck? (if it's me it does not work)

jukkar commented:
> can you request a recheck? (if it's me it does not work)

Recheck has not worked for me for long time. Perhaps it has been removed altogether. You need to rebase and force push, or close and re-open this PR.

tbursztyka commented:
@jukkar ok I thought it was some access rights issue

jukkar commented on include/net/net_if.h:
```diff
@@ -56,7 +56,8 @@ struct net_if_addr {
 
```
In commit message "net/iface: Coalesce all DAD timers through one", the string "Each net_if_addr can be added to the timer handler via a slist, then." sounds a bit too French. What about just saying "Then each net_if_addr can be added to the timer handler via a slist."

jukkar commented on include/net/net_if.h:
```diff
@@ -231,12 +228,19 @@ struct net_if_ipv6 {
+	sys_snode_t rs_node;
 
-	/** IPv6 hop limit */
-	u8_t hop_limit;
```
Why do we move the hop limit in this commit "net/iface: Coalesce all RS timers through one" outside of IPV6_ND? Does this belong to some other commit?

tbursztyka replied:
It never was part of IPV6_ND, only RS. I did move it in order to get the RS-related attributes inside a new ifdef for IPV6_ND. It's fine here.

jukkar commented on include/net/net_if.h:
```diff
@@ -220,9 +220,6 @@ struct net_if_ipv6 {
 	/** Prefixes */
```
Also better would be just to say "Then each net_if_ipv6 can be added to the timer handler via a slist" in commit "net/iface: Coalesce all RS timers through one"

jukkar commented on include/net/net_if.h:
```diff
@@ -1154,6 +1160,28 @@ void net_if_ipv6_prefix_unset_timer(struct net_if_ipv6_prefix *prefix);
+#else
+static inline struct in6_addr *net_if_router_ipv6(struct net_if_router *router)
+{
+	static struct in6_addr addr;
```
Return `net_ipv6_unspecified_address()` here so we save some bytes.

tbursztyka replied:
good point

tbursztyka replied:
actually, this will be nuked out by the compiler (that's being made so code using the function can be enclosed in IS_ENABLED() )

jukkar commented on include/net/net_if.h:
```diff
@@ -1566,6 +1594,28 @@ bool net_if_ipv4_maddr_rm(struct net_if *iface, const struct in_addr *addr);
+#else
+static inline struct in_addr *net_if_router_ipv4(struct net_if_router *router)
+{
+	static struct in_addr addr;
```
Here we should return `net_ipv4_unspecified_address()`

jukkar commented on include/net/net_if.h:
```diff
@@ -1628,6 +1628,17 @@ static inline struct in_addr *net_if_router_ipv4(struct net_if_router *router)
 struct net_if_router *net_if_ipv4_router_lookup(struct net_if *iface,
```
The commit subject of "net/iface: Adding IPv4 router counterpart functions to IPv6" looks wrong (because IPv6 is mentioned). What about just saying "net/iface: Add IPv4 route find and add".

tbursztyka replied:
ok

jukkar reviewed (changes requested):
Getting there, just minor tweaks needed.

jukkar reviewed (approved):
LGTM
